### PR TITLE
Add schedule end date/count field

### LINKED ISSUE #1568: [Feature] Add end date to repeated Schedules
### Verified feature request does not already exist?

- [X] I have searched and found no existing issue

### 💻

- [ ] Would you like to implement this feature?

### Pitch: what problem are you trying to solve?

Hello,

In this screen I searched a long time and figured out it was probably a missing feature.

We can set when the schedule starts but we can't end it by choosing a date.

image

For example; I had to set 23x scheduled entries for all the individual payments for my appliances, which are monthly payments.

### Describe your ideal solution to this problem

Add a date field and date picker to be able to select an ending date. Once the end date is reached the schedule would be moved under "Show completed schedules".

### Teaching and learning

_No response_

### PR DESCRIPTION
**Pull request by jfdoming**
Resurrecting #407! This PR adds an end date/max occurrences field to the schedule creator. Closes #1568 and #241.

<h3>Screenshots</h3>
<h4>Current popup:</h4>
image
<h4>New popup:</h4>
image

### CODE REVIEW

File: upcoming-release-notes/1899.md
Comment (Shazib) [inline upcoming-release-notes/1899.md:2]: I'd tag this as a feature personally!

File: packages/loot-core/src/types/models/schedule.d.ts
Comment (Shazib) [inline packages/loot-core/src/types/models/schedule.d.ts:29]: Would be be possible to only store `endDate` regardless of `endMode`?

So when the user selects '5 occurrences' we work out the _date_ of the 5th occurrence and store that? Just thinking about saving a column but maybe its not worth the effort.

**Review by Shazib — COMMENTED** [review]
Looks good!

File: packages/desktop-client/src/components/select/RecurringSchedulePicker.js
Comment (joel-jeremy) [inline packages/desktop-client/src/components/select/RecurringSchedulePicker.js:335]: Can we use 'indefinitely' here instead?

File: upcoming-release-notes/1899.md
Comment (jfdoming) [reply upcoming-release-notes/1899.md:2]: Oops totally my bad! I was in bugfixing mode in my mind 😆

File: packages/loot-core/src/types/models/schedule.d.ts
Comment (jfdoming) [reply packages/loot-core/src/types/models/schedule.d.ts:29]: I think it makes showing the future UI a lot easier. For example, if you add a schedule and then go edit the rule later, having a dedicated `endOccurrences` field means we can just use that for hydrating the UI. If we don't have that field, we have to have logic like "generate dates until we pass the end date" which could get pretty inefficient if someone inputs a large # of occurrences. Let me know if this makes sense!

File: packages/desktop-client/src/components/select/RecurringSchedulePicker.js
Comment (jfdoming) [reply packages/desktop-client/src/components/select/RecurringSchedulePicker.js:335]: Good idea!

File: packages/loot-core/src/types/models/schedule.d.ts
Comment (Shazib) [reply packages/loot-core/src/types/models/schedule.d.ts:29]: I'm not too fussed either way, Just a suggestion! But I wouldn't necessarily have thought it would hurt perf too badly even for v large dates, particularly as you only see one at a time.

Comment (jfdoming) [reply packages/loot-core/src/types/models/schedule.d.ts:29]: Yeah, thanks for the suggestion! I think I should have been more clear: I'm talking about the rules page which displays the same selector:
image
If we don't store `endOccurrences`, I think we'd have to recalculate it on this page by manually iterating until we hit `endDate`, no?

Comment (Shazib) [reply packages/loot-core/src/types/models/schedule.d.ts:29]: Yeh you're correct; you'd probably have to iterate through the schedule executions till you reach 'end date'. Even then I dont think it would be too slow.

but given what you've got now works I wouldn't worry too much about it!

**Comment by youngcw** [comment]
@shall0pass This will need testing against templates when one or both of us get a chance.

**Comment by youngcw** [comment]
> This will need testing

I tried this really quick and it caused the template script to get into some sort of infinite loop.  Thats likely not a reason to hold this PR back, but will need fixed soon.

**Comment by jfdoming** [comment]
> > This will need testing
> 
> I tried this really quick and it caused the template script to get into some sort of infinite loop. Thats likely not a reason to hold this PR back, but will need fixed soon.

@youngcw is this related to #1917 at all or is that something else?

**Comment by youngcw** [comment]
#1917 is for something else.  The schedule template processor needs updated to handle the schedules ending

**Comment by jfdoming** [comment]
> #1917 is for something else. The schedule template processor needs updated to handle the schedules ending

Got it. Ill try to replicate this later today

**Comment by jfdoming** [comment]
> #1917 is for something else. The schedule template processor needs updated to handle the schedules ending

I think this is fixed now! Semi-related to the previous issue, just needed to add a further fallback to the logic

**Comment by youngcw** [comment]
> I think this is fixed now! Semi-related to the previous issue, just needed to add a further fallback to the logic

Cool it looks to be working for me.  Ill try some more to see if I can break it, but so far so good!

**Comment by jfdoming** [comment]
(sorry, rebased on master due to merge conflicts)

**Review by shall0pass — CHANGES_REQUESTED** [review]

File: packages/loot-core/src/server/budget/goals/goalsSchedule.ts
Comment (jfdoming) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: > I don't think this works correctly. If you have a schedule that has two occurrences in the same month, the function could exit prematurely. I think diffDays needs to be reevaluated before doing this check.

Oh, nice catch! This should definitely have been reevaluated. I decided to slightly update how the logic works to be more straightforward/correct in general, let me know your thoughts! Also fixed another bug I found incidentally, that one can move to a separate PR if you like.

Comment (youngcw) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: My test budget is crashing with the current build.  Here it is if you want to try and debug it.  I believe it was working before these most recent changes, so its likely somewhere in there.
[template_test_export.zip](https://github.com/actualbudget/actual/files/13547319/template_test_export.zip)

Comment (jfdoming) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: Ah yikes, sorry about that. I'll look into this later today.l and add what I find to my own manual tests.

Comment (jfdoming) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: Alright, sorry for the pain here. I was able to run your test budget with no issues in the latest commit.

**Comment by youngcw** [comment]
Looks to be working with the template test file!  I tried both number of occurrences and end date and it seems to work right.  @shall0pass You understand the schedule template a lot better than I do so Ill defer reviewing the code to you.

File: packages/loot-core/src/server/budget/goals/goalsSchedule.ts
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: I've attached a test budget.
[2023-12-05-My-Finances-564f13f.zip](https://github.com/actualbudget/actual/files/13573272/2023-12-05-My-Finances-564f13f.zip)

In this test file if you use the "Savings" category that has a schedule called "breakWeekend", the months of December and March aren't right.  Those days both fall on a Saturday those months and only budget for one of the two schedules.  
I haven't tried it, but I think this only occurs on the 'before' style schedules.  What if instead of passing the weekendSolveMode variable of the actual schedule, we force it to 'after' so diffDays won't automatically be evaluated as 0 and will then look for a second schedule for the month?  Thanks for all the cleanup you've done!

Comment (jfdoming) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: Oh, nice catch! This was just one place I missed in my refactor, I believe this should be fixed in that budget now.

**Comment by youngcw** [comment]
Fixes #506 as well

File: packages/loot-core/src/server/budget/goals/goalsSchedule.ts
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goals/goalsSchedule.ts]: I think you've got it!

**Review by shall0pass — DISMISSED** [review]
I'm good with all of the template interactions.  I'll wait on @Shazib or @joel-jeremy for their reviews and final approval.

**Comment by jfdoming** [comment]
(just rebased on master again due to merge conflicts, and fixed one let -> const)

**Review by joel-jeremy — APPROVED** [review]
Looks good!

**Comment by jfdoming** [comment]
> Looks good!

Awesome, thanks! Are we good to merge, then? (I don't have merge permissions, not sure how this normally works)

**Review by shall0pass — APPROVED** [review]

**Comment by Kidglove57** [comment]
I have two instances of Actual, one updated today in “edge” and the other using the “latest” tag. In addition I had four scheduled payments auto posting this morning. They are set up for “after the weekend” but that was not relevant today as it was a Friday. This is what I observed:
1) These schedules posted automatically and correctly in “latest”.
2) However, in my experimental file using “edge” none of the scheduled payments auto posted. Instead they simply showed as “due”.

It may be a coincidence but the merge of the above PR seems to be the difference between the behaviour on my two files. I wonder therefore if there may still be a bug here to be hunted down?

**Comment by carkom** [comment]
@jfdoming 

Yes, I see the same as @Kidglove57 . Any schedules that were created before this PR are given a "null" for "endMode" which means it doesn't run. Is there a catch for this case in the code? I haven't read through it, so it may already exist, but something like if "isRecurring === true" and "endMode == null" then assume it's indefinite. Sorry again if you've already got this in there but it may not be working properly.

**Comment by jfdoming** [comment]
> @jfdoming
> 
> Yes, I see the same as @Kidglove57 . Any schedules that were created before this PR are given a "null" for "endMode" which means it doesn't run. Is there a catch for this case in the code? I haven't read through it, so it may already exist, but something like if "isRecurring === true" and "endMode == null" then assume it's indefinite. Sorry again if you've already got this in there but it may not be working properly.

Yikes, sorry about that. I thought I handled those cases but clearly not. I'll look into it now

**Comment by jfdoming** [comment]
Hmm, @Kidglove57 I'm struggling to reproduce this issue. Do you have a redacted sample budget you can share where this occurs? I tried making a schedule with the endMode initially `undefined`, but that seemed to work fine

**Comment by Kidglove57** [comment]
I’ve got two more scheduled transactions due to post tomorrow (Sunday). They are useful for this trial as, unlike the previous failed posts, these are without “after the weekend” enabled. I’ll see what happens and if they auto post and report back here.

**Comment by jfdoming** [comment]
> I’ve got two more scheduled transactions due to post tomorrow (Sunday). They are useful for this trial as, unlike the previous failed posts, these are without “after the weekend” enabled. I’ll see what happens and if they auto post and report back here.
> 
> The problem with test budgets is that the only way to reproduce this is when real future scheduled transactions are due to post. I will add in a brand new test schedule as a further trial.

I was currently testing things by altering my system time. So I think I would be able to fast forward things that way. But maybe I'm misunderstanding! At any rate, awaiting your test results 🙏

**Comment by Kidglove57** [comment]
Thank you - I will. Bearing in mind @carkom comment, and as an extra check, may I ask if the schedules you were testing were created prior to this release?

**Comment by Kidglove57** [comment]
Good morning. I am glad to bring good news. Both my two pre existing scheduled transactions and my test transaction posted successfully this morning (Sunday) using the latest Edge. I have no way of assessing why all four scheduled payments due on Friday failed (unless it was the fact that they were marked “after the weekend”). I’ll set up another test transaction today with the “after the weekend” marker and see what happens on Monday. Fingers crossed!

**Comment by Kidglove57** [comment]
All posted successfully this morning! I’ll keep an eye on this and thank you for your patience. Hopefully my experience on Friday was a one off!

**Comment by jfdoming** [comment]
> All posted successfully this morning! I’ll keep an eye on this and thank you for your patience. Hopefully my experience on Friday was a one off!

Glad to hear it worked! It is definitely strange that it didn't work on Friday though... Maybe something with "after the weekend" on a Friday. I'll keep following this thread just in case!

**Pull request merged**
